pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,308 @@
+	if !controllerutil.ContainsFinalizer(extension, operatorv1alpha1.FinalizerName) {
+		log.Info("Adding finalizer")
+		if err := controllerutils.AddFinalizers(ctx, r.RuntimeClient, extension, operatorv1alpha1.FinalizerName); err != nil {
+			return fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	log.Info("Reconciling extension virtual resources")
+	if err := r.reconcileVirtualClusterResources(ctx, log, virtualClusterClient, extension); err != nil {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionReconcileFailed, err.Error())
+		return errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions))
+	}
+
+	conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionTrue, ConditionReconcileSuccess, fmt.Sprintf("Extension %q has been reconciled successfully", extension.Name))
+	return r.updateExtensionStatus(ctx, log, extension, conditions)
+}
+
+func (r *Reconciler) reconcileVirtualClusterResources(ctx context.Context, log logr.Logger, virtualClusterClient client.Client, extension *operatorv1alpha1.Extension) error {
+	// return early if we do not have to make a deployment
+	if extension.Spec.Deployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment.Helm == nil {
+		return r.deleteVirtualClusterResources(ctx, log, virtualClusterClient, extension)
+	}
+
+	if err := r.reconcileControllerDeployment(ctx, virtualClusterClient, extension); err != nil {
+		return fmt.Errorf("failed to reconciler controller deployment: %w", err)
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerDeployment applied successfully")
+
+	if err := r.reconcileControllerRegistration(ctx, virtualClusterClient, extension); err != nil {
+		return fmt.Errorf("failed to reconciler controller registration: %w", err)
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerRegistration applied successfully")
+	return nil
+}
+
+func (r *Reconciler) reconcileControllerDeployment(ctx context.Context, virtualClusterClient client.Client, extension *operatorv1alpha1.Extension) error {
+	controllerDeployment := &gardencorev1.ControllerDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension.Name,
+		},
+	}
+
+	if _, err := controllerutil.CreateOrPatch(ctx, virtualClusterClient, controllerDeployment,
+		func() error {
+			controllerDeployment.Helm = &gardencorev1.HelmControllerDeployment{
+				Values:        extension.Spec.Deployment.ExtensionDeployment.Values,
+				OCIRepository: extension.Spec.Deployment.ExtensionDeployment.Helm.OCIRepository,
+			}
+			return nil
+		}); err != nil {
+		return fmt.Errorf("failed to create or update ControllerDeployment: %w", err)
```

Comment:
```suggestion
		return fmt.Errorf("failed to reconcile ControllerDeployment: %w", err)
```

above, you wrote `controller registration`, but here you write `ControllerDeployment`